tree/dataframe/inc/ROOT/RDF/ColumnReaders.hxx

```diff
@@ -0,0 +1,331 @@
+   /// in contiguous memory.
+   EStorageType fStorageType = EStorageType::kUnknown;
+
+   /// We return a reference to this RVec to clients, to guarantee a stable address and contiguous memory layout.
+   RVec<T> fRVec;
+
+   /// Whether we already printed a warning about performing a copy of the TTreeReaderArray contents
+   bool fCopyWarningPrinted = false;
+
+public:
+   RTreeColumnReader(TTreeReader &r, const std::string &colName)
+      : fTreeArray(std::make_unique<TTreeReaderArray<T>>(r, colName.c_str()))
+   {
+   }
+
+   RVec<T> &Get(Long64_t) final
+   {
+      auto &readerArray = *fTreeArray;
+      // We only use TTreeReaderArrays to read columns that users flagged as type `RVec`, so we need to check
+      // that the branch stores the array as contiguous memory that we can actually wrap in an `RVec`.
+      // Currently we need the first entry to have been loaded to perform the check
+      // TODO Move check to constructor once ROOT-10823 is fixed and TTreeReaderArray itself exposes this information
+      const auto arrSize = readerArray.GetSize();
+      if (EStorageType::kUnknown == fStorageType && arrSize > 1) {
+         // We can decide since the array is long enough
+         fStorageType = EStorageType::kContiguous;
+         for (auto i = 0u; i < arrSize - 1; ++i) {
+            if ((char *)&readerArray[i + 1] - (char *)&readerArray[i] != sizeof(T)) {
+               fStorageType = EStorageType::kSparse;
+               break;
+            }
+         }
+      }
+
+      const auto readerArraySize = readerArray.GetSize();
+      if (EStorageType::kContiguous == fStorageType ||
+          (EStorageType::kUnknown == fStorageType && readerArray.GetSize() < 2)) {
+         if (readerArraySize > 0) {
+            // trigger loading of the contents of the TTreeReaderArray
+            // the address of the first element in the reader array is not necessarily equal to
+            // the address returned by the GetAddress method
+            auto readerArrayAddr = &readerArray.At(0);
+            RVec<T> rvec(readerArrayAddr, readerArraySize);
+            std::swap(fRVec, rvec);
+         } else {
+            RVec<T> emptyVec{};
+            std::swap(fRVec, emptyVec);
+         }
+      } else {
+         // The storage is not contiguous or we don't know yet: we cannot but copy into the rvec
+#ifndef NDEBUG
+         if (!fCopyWarningPrinted) {
+            Warning("RColumnValue::Get",
+                    "Branch %s hangs from a non-split branch. A copy is being performed in order "
+                    "to properly read the content.",
+                    readerArray.GetBranchName());
+            fCopyWarningPrinted = true;
+         }
+#else
+         (void)fCopyWarningPrinted;
+#endif
+         if (readerArraySize > 0) {
+            RVec<T> rvec(readerArray.begin(), readerArray.end());
+            std::swap(fRVec, rvec);
```

Comment:
No idea if this makes any difference in terms of optimisations that can be applied.
```suggestion
            fRVec.assign(readerArray.begin(), readerArray.end());
```

Reply:
`fRVec` might not have the right size

Reply:
It will resize if necessary, but changing it won't probably gain anything. Like this, the visual impression of the four cases is the same, so it's easy to read. 👍

Reply:
ah, I did not realize `assign` would resize as needed. uhm then a reader might be surprised by the `swap`. I'll add your suggestion